rust/core-lib/src/movement.rs

```diff
@@ -88,14 +84,84 @@ fn vertical_motion(
     } else {
         line.saturating_add(line_delta as usize)
     };
-    let n_lines = view.line_of_offset(text, text.len());
     if line > n_lines {
         return (text.len(), Some(col));
     }
     let new_offset = view.line_col_to_offset(text, line, col);
     (new_offset, Some(col))
 }
 
+/// Compute movement based on vertical motion by the given number of lines skipping
+/// any line that is shorter than the current cursor position.
+fn vertical_motion_exact_pos(
+    r: SelRegion,
+    view: &View,
+    text: &Rope,
+    line_delta: isize,
+    modify: bool,
+) -> (usize, Option<HorizPos>) {
+    let (col, init_line, n_lines) = selection_position(r, view, text, line_delta, modify);
+
+    let mut line_length = view.offset_of_line(text, init_line.saturating_add(1))
+        - view.offset_of_line(text, init_line);
+    if line_delta < 0 && (-line_delta as usize) > init_line {
+        return (view.line_col_to_offset(text, init_line, col), Some(col));
+    }
+    let mut line = if line_delta < 0 {
+        init_line - (-line_delta as usize)
+    } else {
+        init_line.saturating_add(line_delta as usize)
+    };
+
+    // If the active columns is longer than the current line, use the current line length.
+    let col = if line_length < col { line_length - 1 } else { col };
+
+    loop {
+        line_length =
```

Comment:
although I think it's fine, I did stare at this for a while, thinking about the EOF case. I think the `saturating_add` is unnecessary here, because line + 1 for any valid line is <= the total number of bytes in the document; if this number would overflow we've already crashed.

Reply:
I'll defer to your expertise here, will change.